### ISSUE TITLE
Inconsistent CS8602 if the return value of a method is stored in a variable

### ISSUE THREAD

**Issue by Foxite** [dotnet/roslyn#39749]
**Version Used**: 3.3.1-beta3-19461-02 (2fd12c21)

**Steps to Reproduce**:

```csharp
#nullable enable
public class TestClass {
	private bool TryGetValue([NotNullWhen(true)] out string? value) {
		value = "Bla";
		return true;
	}

	public void DoStuff1() {
		bool success = TryGetValue(out string? result);
		if (success) {
			Console.WriteLine(result.Length);
		}
	}

	public void DoStuff2() {
		if (TryGetValue(out string? result)) {
			Console.WriteLine(result.Length);
		}
	}
}
```

**Expected Behavior**: `DoStuff1` and `DoStuff2` do not generate different warnings (better yet, neither generates a warning)

**Actual Behavior**: A CS8602 warning is generated in `DoStuff1`, saying `result` may be null inside the if statement, which is incorrect. `DoStuff2` does not have such a warning.

**Comment by gafter** [comment]
Flow annotations such as `[NotNullWhen(true)]` affect the null-when-true and null-when-false state of the out variable based on the return value of the method.  But such bool-value-dependent states are not preserved when stored in `bool` variables.  That is by design.
